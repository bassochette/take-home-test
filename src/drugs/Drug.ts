export class Drug {
    name: string;
    expiresIn: any;
    benefit: any;

    constructor(name, expiresIn, benefit) {
        this.name = name;
        this.expiresIn = expiresIn;
        this.benefit = benefit;
    }
}