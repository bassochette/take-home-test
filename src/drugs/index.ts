export * from './Drug';
export * from './Dafalgan/Daflagan';
export * from './Doliprane/Doliprane';
export * from './HerbalTea/HerbalTea';
export * from './Fervex/Fervex';
export * from './MagicPill/MagicPill';
