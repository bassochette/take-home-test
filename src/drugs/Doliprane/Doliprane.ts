import { Drug } from '../Drug';

export class Doliprane extends Drug {
  constructor(expiresIn = 20, benefit = 30) {
    super('Doliprane', expiresIn, benefit);
  }
}
