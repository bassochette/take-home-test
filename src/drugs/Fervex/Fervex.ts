import { Drug } from '../Drug';

export class Fervex extends Drug {
  constructor(expiresIn = 5, benefit = 40) {
    super('Fervex', expiresIn, benefit);
  }

  updateBenefit() {
    this.expiresIn--;
    if (this.expiresIn < 0) {
      this.benefit = 0;
    } else if (this.expiresIn <= 5) {
      this.increaseBenefit(3);
    } else if (this.expiresIn <= 10) {
      this.increaseBenefit(2);
    }
  }
}
