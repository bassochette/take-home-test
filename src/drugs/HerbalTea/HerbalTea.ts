import { Drug } from '../Drug';

export class HerbalTea extends Drug {
  constructor(expiresIn = 10, benefit = 5) {
    super('Herbal Tea', expiresIn, benefit);
  }

  updateBenefit() {
    this.expiresIn--;
    if (this.expiresIn < 0) {
      this.increaseBenefit(2);
    } else {
      this.increaseBenefit();
    }
  }
}
