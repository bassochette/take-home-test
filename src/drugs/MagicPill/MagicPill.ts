import { Drug } from '../Drug';

export class MagicPill extends Drug {
  constructor(expiresIn = 15, benefit = 40) {
    super('Magic Pill', expiresIn, benefit);
  }

  updateBenefit() {
    // Does not expires nor decrease in benefit;
  }
}
