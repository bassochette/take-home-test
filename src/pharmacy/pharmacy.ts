import { Drug } from '../drugs';

export class Pharmacy {
  drugs: Drug[];
  constructor(drugs = []) {
    this.drugs = drugs;
  }
  updateBenefitValue(): Drug[] {
    for (const drug of this.drugs) {
      drug.updateBenefit();
    }

    return this.drugs;
  }
}
