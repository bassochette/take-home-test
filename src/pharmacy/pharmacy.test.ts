import { Pharmacy } from './pharmacy';
import { Drug } from '../drugs/Drug';
import {Doliprane} from "../drugs/Doliprane/Doliprane";
import {HerbalTea} from "../drugs/HerbalTea/HerbalTea";
import {Fervex} from "../drugs/Fervex/Fervex";
import {MagicPill} from "../drugs/MagicPill/MagicPill";

describe('Pharmacy', () => {
  it('should decrease the benefit and expiresIn', () => {
    expect(
      new Pharmacy([new Drug('test', 2, 3)]).updateBenefitValue(),
    ).toEqual([new Drug('test', 1, 2)]);
  });
  it('match snapshot', () => {
    const drugs = [
      new Doliprane(),
      new HerbalTea(),
      new Fervex(),
      new MagicPill(),
    ];

    const pharmacy = new Pharmacy(drugs);
    for (let elapsedDays = 0; elapsedDays < 30; elapsedDays++) {
      pharmacy.updateBenefitValue();
    }
    expect(pharmacy).toMatchSnapshot();
  });
});
