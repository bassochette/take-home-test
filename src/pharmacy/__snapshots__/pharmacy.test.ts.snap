// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pharmacy match snapshot 1`] = `
Pharmacy {
  "drugs": Array [
    Doliprane {
      "benefit": 0,
      "expiresIn": -10,
      "name": "Doliprane",
    },
    HerbalTea {
      "benefit": 50,
      "expiresIn": -20,
      "name": "Herbal Tea",
    },
    Fervex {
      "benefit": 0,
      "expiresIn": -25,
      "name": "Fervex",
    },
    MagicPill {
      "benefit": 40,
      "expiresIn": 15,
      "name": "Magic Pill",
    },
  ],
}
`;
